torch_xla/csrc/xla_sharding_util.cpp

```diff
@@ -457,9 +482,18 @@ std::vector<at::Tensor> ShardingUtil::ShardTensor(
     XLA_CHECK(sharding.tile_shape().dimensions_size() <= 2);
     XLA_CHECK(tensor.sizes().size() >= sharding.tile_shape().dimensions_size());
 
-    auto shard_shape = GetShardShape(tensor, sharding);
-    auto shard_indices = GetShardIndicesForDevices(
-        shard_shape, tensor.sizes().vec(), sharding, devices);
+    auto shard_shape = GetShardShape(tensor.sizes().vec(), sharding);
+    if (minibatch) {
+      shard_shape[0] = tensor.sizes().vec()[0] / devices.size();
```

Comment:
Can we call `GetShardShape` with the global shape instead of modifying the shard shape here? I think you'd just need to change the `sharding` parameter from `xla::OpSharding` to `ShardingSpecPtr`, then the global shape can come from the ShardingSpec,

We could then also get rid of the `minibatch` parameter